draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,190 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+a terminal state is reached.
+
+Let us call the initiating party the "Leader" and the responding party the
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), the Leader sends its prep share
+  to the Helper, who computes the prep message locally (the computations are
+  specified in the following sections), computes its output share, then sends
+  the prep message to the Leader. Preparation requires just one request from
+  the Leader to the Helper.
+
+* For a 2-round VDAF (e.g., Poplar1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two requests.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+The message types are as follows:
+
+* `InitializeMessage(prep_share)` - Sent by the Leader to the Helper in the
+  initial request. The payload includes its first-round prep share.
+
+* `ContinueMessage(prep_msg, prep_share)` - Sent by the Leader or Helper in an
+  intermediate request or response. The payload includes the prep message from
+  the current round and the sender's prep share for the next round.
+
+* `FinishMessage(prep_msg)` - Sent by the Leader in the last request
+  (even-round VDAF) or the Helper in the last response (odd-round VDAF). The
+  payload is the final-round prep message.
+
+Each Aggregator handles these messages according to the following transition rule:
+
+~~~ transition
+def ping_pong_next(Vdaf,
+                   is_leader: bool,
+                   agg_param: Vdaf.AggParam,
+                   state: State,
+                   inbound: Optional[Message],
+                   ) -> (State, Optional[Message]):
+    if type(inbound) == InitializeMessage:
+        prep_msg, peer_prep_share = None, inbound.prep_share
+    elif type(inbound) == ContinueMessage:
+        prep_msg, peer_prep_share = inbound.prep_msg, inbound.prep_share
+    elif type(inbound) == FinishMessage:
+        prep_msg, peer_prep_share = inbound.prep_msg, None
+    else: # inbound == None
```

Comment:
What do you think about changing the `None` case to be an explicit `FinishedMessage`? This would mirror what's in DAP, and would make it much clearer here which case is being handled. In practice, any request/response protocol will need to send _some_ message, so I think it makes some sense to define what that message is as `FinishedMessage`. (name is up for debate, just suggesting what we went with in the analogous scenario in DAP)

Reply:
See this comment: https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/240#discussion_r1204848392. I'll add it back.

Reply:
Done!